store.go

```diff
@@ -1,51 +1,70 @@
 package vstore
 
-// Store is the main redux-like store. Use it to access the State properties
-// through the ComponentConstructor.
-type Store struct {
-	r    Reducer
-	mws  []Middleware
-	subs map[*SubFunc]bool
+import (
+	"github.com/gopherjs/vecty"
+	"sync"
+)
+
+// store is the main redux-like store. Use it to access the State properties.
+type Store interface {
+	Connect(comp StoreComponent) StoreComponent
+	Dispatch(action interface{})
+	State() interface{}
 }
 
-// SubFunc is a subscription callback.
-type SubFunc func(s *Store)
+// store satisfies Store.
+type store struct {
+	r   Reducer
+	mws []Middleware
+
+	mtx   sync.RWMutex
+	comps map[*storeComponent]bool
+}
 
 // Middleware is a simple middleware that gets called before any Reduce calls.
 type Middleware func(action interface{})
 
-// CreateStore returns a store that you can make global dispatches to in order
+// New returns a store that you can make global dispatches to in order
 // to update the passed global state.
-func CreateStore(r Reducer, mws ...Middleware) *Store {
-	s := Store{r: r, mws: mws, subs: map[*SubFunc]bool{}}
-
-	return &s
+func New(r Reducer, mws ...Middleware) Store {
+	return &store{r: r, mws: mws, comps: map[*storeComponent]bool{}}
 }
 
 // Dispatch takes an action and passes it to all middlewares and the Reduce function.
-// It also calls any subscription functions if exist.
-func (s *Store) Dispatch(action interface{}) {
+// NOTE: vecty.Rerender is called in a goroutine for each subscribed component.
+func (s *store) Dispatch(action interface{}) {
 	for _, m := range s.mws {
 		m(action)
 	}
 
 	s.r.Reduce(action)
 
-	for sub := range s.subs {
-		(*sub)(s)
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	for comp := range s.comps {
+		comp := comp
```

Comment:
unnecessary reassign?

Reply:
It actually creates a new var here, since for range loops reuse the same pointer address per iteration. See parallelized table driven tests https://rakyll.org/parallelize-test-tables/ for details.

Reply:
@norunners this is because `t.Run` does not take an extra parameter to register the changing loop variable. In our case, we are passing `comp` to `go vecty.Rerender()`. Which means, that `comp` is being registered because we are passing it as a function parameter and not as a closure. So this should be safe to remove. Check this wiki for more info: https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables